Wheel event does not stop macOS smooth keyboard scroll

rr-codes opened the pull request:
#### 8c822fa15adbdecd22d0c3ece07c9e9aa2260af1
<pre>
Wheel event does not stop macOS smooth keyboard scroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=244303">https://bugs.webkit.org/show_bug.cgi?id=244303</a>
rdar://99101109

Reviewed by Tim Horton.

Currently, if you enable EventHandlerDrivenSmoothKeyboardScrolling,
scroll using the keyboard, then immediately after use the trackpad to
scroll, the webpage will scroll back to the position where the keyboard
scroll left off. This patch improves the smooth scroll setting so that a
wheel event tells the KeyboardScrollingAnimator to stop scrolling.

This adds a method to KeyboardScrollingAnimator, stopScrollingImmediately
 that can be called to stop the animation without taking the time
To decelerate, which is what should happen when a wheel event
Interrupts the keyboard scroll.

* LayoutTests/fast/scrolling/mac/keyboard-scrolling-with-mouse-scroll.html:
* Source/WebCore/platform/KeyboardScrollingAnimator.cpp
(WebCore::KeyboardScrollingAnimator::stopScrollingImmediately):
* Source/WebCore/platform/KeyboardScrollingAnimator.h
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp

Canonical link: <a href="https://commits.webkit.org/254561@main">https://commits.webkit.org/254561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25435827d6e73b37d5a9daa6912d7d5f8373b64a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89330 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98665 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154971 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93338 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32393 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27916 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81694 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93090 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25733 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76272 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25675 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80609 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80553 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68652 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30167 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14615 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29894 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15564 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3201 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33342 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38521 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32046 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34655 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -361,15 +361,17 @@ void FocusController::setFocusedFrame(Frame* frame)
     // Now that the frame is updated, fire events and update the selection focused states of both frames.
     if (auto* oldFrameView = oldFrame ? oldFrame->view() : nullptr) {
-        oldFrameView->stopKeyboardScrollAnimation();
+        oldFrameView->stopKeyboardScrollAnimation([] {
```
should the below lines go in this lambda block?

rr-codes commented on Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
```diff
@@ -339,13 +341,27 @@ void KeyboardScrollingAnimator::stopKeyboardScrollAnimation()
     if (!m_scrollTriggeringKeyIsPressed)
         return;
 
     stopKeyboardScrollAnimation();
```
should this be called here? or below line 357? or somewhere else entirely?

rr-codes replied:
without it, scrolling never ceases, so I put it here without knowing any other unintended consequences

rr-codes commented on Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
```diff
@@ -104,6 +105,16 @@ void EventDispatcher::initializeConnection(IPC::Connection* connection)
 {
     auto processingSteps = OptionSet<WebCore::WheelEventProcessingSteps> { WheelEventProcessingSteps::MainThreadForScrolling, WheelEventProcessingSteps::MainThreadForBlockingDOMEventDispatch };
+
+    ensureOnMainRunLoop([pageID] {
```
not entirely sure about this but it was suggested in the superseded PR

hortont424 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -361,15 +361,17 @@ void FocusController::setFocusedFrame(Frame* frame)
     // Now that the frame is updated, fire events and update the selection focused states of both frames.
     if (auto* oldFrameView = oldFrame ? oldFrame->view() : nullptr) {
-        oldFrameView->stopKeyboardScrollAnimation();
+        oldFrameView->stopKeyboardScrollAnimation([] {
```
Smush onto one line (or maybe make it take optional/null/something so we don't have to pass an empty block?)

rr-codes replied:
should the lines below this actually go into this completion block instead? Or is it fine as-is?

hortont424 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2494,16 +2494,20 @@ void Page::stopKeyboardScrollAnimation()
+                if (!page)
+                    return;
+                if (auto scrollingCoordinator = frame->page()->scrollingCoordinator())
+                    scrollingCoordinator->setKeyboardScrollAnimationInProgress(*scrollableArea, false);
```
This seems racy -- what if you've started another scroll in the meantime?

rr-codes replied:
Ah ok. Do you have any ideas on how to make it safer?

hortont424 replied:
We usually use incrementing integer request identifiers ("make sure I'm still the latest request to stop, if somebody else came in in the meantime then do nothing")

But it's possible that this thing would be better handled by tokens ("everyone who is doing a keyboard scroll takes out a refcounted token and drops it when the scroll ends, and if there are any outstanding tokens then KeyboardScrollAnimationInProgress == true, and when the last outstanding one dies do whatever you need").

I'm not sure :) Would have to peruse this code a bit more.

hortont424 reviewed (commented):
Needs a test. But otherwise, yay, much simpler!

pxlcoder commented on LayoutTests/fast/scrolling/mac/keyboard-scrolling-with-mouse-scroll.html:
```diff
@@ -0,0 +1,44 @@
+                await UIHelper.mouseWheelScrollAt(10, 10, 0, 1, 0, 10);
+
+                const position = window.scrollY;
+                if (position <= 1) 
+                    debug("Successful.");
+                else
+                    debug("Unsuccessful. window.scrollY == " + position);
```
This can use `shouldBeLessThanOrEqual` from `js-test`.

pxlcoder commented on LayoutTests/fast/scrolling/mac/keyboard-scrolling-with-mouse-scroll.html:
```diff
@@ -0,0 +1,44 @@
+
+<head>
+    <script src="../../../resources/ui-helper.js"></script>
+    <script src="../../../resources/js-test-pre.js"></script>
```
You can use `../../../resources/js-test.js` instead of the `pre`/`post` versions.

rr-codes committed:
Wheel event does not stop macOS smooth keyboard scroll
https://bugs.webkit.org/show_bug.cgi?id=244303
rdar://99101109

Reviewed by Tim Horton.

Currently, if you enable EventHandlerDrivenSmoothKeyboardScrolling,
scroll using the keyboard, then immediately after use the trackpad to
scroll, the webpage will scroll back to the position where the keyboard
scroll left off. This patch improves the smooth scroll setting so that a
wheel event tells the KeyboardScrollingAnimator to stop scrolling.

This adds a method to KeyboardScrollingAnimator, stopScrollingImmediately
 that can be called to stop the animation without taking the time
To decelerate, which is what should happen when a wheel event
Interrupts the keyboard scroll.

* LayoutTests/fast/scrolling/mac/keyboard-scrolling-with-mouse-scroll.html:
* Source/WebCore/platform/KeyboardScrollingAnimator.cpp
(WebCore::KeyboardScrollingAnimator::stopScrollingImmediately):
* Source/WebCore/platform/KeyboardScrollingAnimator.h
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp

Canonical link: https://commits.webkit.org/254561@main
